### PR TITLE
CB-19007 Include conclusion checker into cluster stop flow

### PR DESCRIPTION
**Pull request by lturcsanyi**
See detailed description in the commit message.

### CODE REVIEW

File: core/src/main/java/com/sequenceiq/cloudbreak/reactor/handler/cluster/ClusterStopFailedHandler.java [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+    }
+
+    @Override
+    protected Selectable doAccept(HandlerEvent<ClusterStopFailedRequest> event) {
+        ClusterStopFailedRequest request = event.getData();
+        LOGGER.info("Handle ClusterStopFailedRequest, stackId: {}", request.getResourceId());
+        try {
+            ConclusionChecker conclusionChecker = conclusionCheckerFactory.getConclusionChecker(ConclusionCheckerType.DEFAULT);
+            ConclusionResult conclusionResult = conclusionChecker.doCheck(request.getResourceId());
+            if (entitlementService.conclusionCheckerSendUserEventEnabled(ThreadBasedUserCrnProvider.getAccountId()) && conclusionResult.isFailureFound()) {
```
Comment (sodre90) [inline core/src/main/java/com/sequenceiq/cloudbreak/reactor/handler/cluster/ClusterStopFailedHandler.java:60]: wouldn't be better to move the event sending to the end of conclusionChecker.doCheck method?

**Review by sodre90 — CHANGES_REQUESTED** [review]

**Review by sodre90 — APPROVED** [review]

File: core/src/main/java/com/sequenceiq/cloudbreak/conclusion/ConclusionCheckerService.java [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+    @Inject
+    private EntitlementService entitlementService;
+
+    public void runConclusionChecker(Long resourceId, String eventType, ResourceEvent resourceEvent, String... eventMessageArgs) {
+        try {
+            ConclusionChecker conclusionChecker = conclusionCheckerFactory.getConclusionChecker(ConclusionCheckerType.DEFAULT);
+            ConclusionResult conclusionResult = conclusionChecker.doCheck(resourceId);
+            if (entitlementService.conclusionCheckerSendUserEventEnabled(ThreadBasedUserCrnProvider.getAccountId()) && conclusionResult.isFailureFound()) {
+                flowMessageService.fireEventAndLog(resourceId, eventType, resourceEvent, getFullEventMessageArgs(eventMessageArgs, conclusionResult));
+            }
```
Comment (gregito) [inline core/src/main/java/com/sequenceiq/cloudbreak/conclusion/ConclusionCheckerService.java:38]: a bit out of context but do we really need to execute the check when for example the entitlement is not granted and we are not able to fire any event due to that?

File: core/src/test/java/com/sequenceiq/cloudbreak/conclusion/ConclusionCheckerServiceTest.java [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+                ? new ConclusionResult(List.of(Conclusion.failed("error", "details", ConclusionStep.class)))
+                : new ConclusionResult(List.of());
+        when(conclusionChecker.doCheck(eq(STACK_ID))).thenReturn(conclusionResult);
+        when(conclusionCheckerFactory.getConclusionChecker(eq(ConclusionCheckerType.DEFAULT))).thenReturn(conclusionChecker);
+        when(entitlementService.conclusionCheckerSendUserEventEnabled(anyString())).thenReturn(Boolean.TRUE);
+
+        ThreadBasedUserCrnProvider.doAs(TEST_USER_CRN,
+                () -> underTest.runConclusionChecker(STACK_ID, Status.STOP_FAILED.name(), ResourceEvent.CLUSTER_STOP_FAILED, "eventMessageArg"));
+
+        verify(conclusionChecker, times(1)).doCheck(eq(STACK_ID));
```
Comment (gregito) [inline core/src/test/java/com/sequenceiq/cloudbreak/conclusion/ConclusionCheckerServiceTest.java:62]: please verify that as well there is no further interactions with this mock

File: core/src/test/java/com/sequenceiq/cloudbreak/conclusion/ConclusionCheckerServiceTest.java [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+        VerificationMode verificationMode = conclusionFound ? times(1) : never();
+        verify(flowMessageService, verificationMode).fireEventAndLog(eq(STACK_ID), eq(Status.STOP_FAILED.name()), eq(ResourceEvent.CLUSTER_STOP_FAILED),
+                eq("eventMessageArg"), eq("[error]"));
+    }
```
Comment (gregito) [inline core/src/test/java/com/sequenceiq/cloudbreak/conclusion/ConclusionCheckerServiceTest.java:66]: also, please add a test that goes into the catch clause even if it is unlikely to happen

**Review by gregito — COMMENTED** [review]
some minor questions

File: core/src/main/java/com/sequenceiq/cloudbreak/conclusion/ConclusionCheckerService.java (hunk 2 above)
Comment (lturcsanyi) [reply core/src/main/java/com/sequenceiq/cloudbreak/conclusion/ConclusionCheckerService.java:38]: Yes, the entitlement is related only to the event sending, the checker will log the conclusion results anyway.

File: core/src/test/java/com/sequenceiq/cloudbreak/conclusion/ConclusionCheckerServiceTest.java (hunk 3 above)
Comment (lturcsanyi) [reply core/src/test/java/com/sequenceiq/cloudbreak/conclusion/ConclusionCheckerServiceTest.java:66]: verify added

File: core/src/test/java/com/sequenceiq/cloudbreak/conclusion/ConclusionCheckerServiceTest.java (hunk 4 above)
Comment (lturcsanyi) [reply core/src/test/java/com/sequenceiq/cloudbreak/conclusion/ConclusionCheckerServiceTest.java]: test added

**Pull request merged**